lite/verifying/provider.go

```diff
@@ -0,0 +1,440 @@
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return vp, nil
+}
+
+// getTrustCommit returns a commit trusted with weak subjectivity. It either:
+// 1. Fetches a commit at height provided in options and ensures the specified commit
+// 	is within the trust period of latest block
+// 2. Trusts the remote node and gets the latest commit
+// 3. Returns an error if the height provided in trust option is too old to sync to latest.
+func getTrustCommit(client lclient.SignStatusClient, options TrustOptions) (types.SignedHeader, error) {
+
+	// Get the lastest commit always
+	latestBlock, err := client.Commit(nil)
+	if err != nil {
+		return types.SignedHeader{}, err
+	}
+
+	// If the user has set a root of trust, confirm it then update to newest
+	if options.TrustHeight != 0 && options.TrustHash != nil {
+		trustBlock, err := client.Commit(&options.TrustHeight)
+		if err != nil {
+			return types.SignedHeader{}, err
+		}
+
+		if latestBlock.Time.Sub(trustBlock.Time) > options.TrustPeriod {
+			return types.SignedHeader{}, fmt.Errorf("Your Trusted Block Height is older than the trust period from Latest Block")
+		}
+
+		trustCommit := trustBlock.SignedHeader
+		if !bytes.Equal(trustCommit.Hash(), options.TrustHash) {
+			return types.SignedHeader{}, fmt.Errorf("WARNING!!! Expected height/hash %v/%X but got %X",
+				options.TrustHeight, options.TrustHash, trustCommit.Hash())
+		}
+		return trustCommit, nil
+	} else {
+
+		latestCommit := latestBlock.SignedHeader
+
+		// NOTE: This should really belong in the callback.
+		// WARN THE USER IN ALL CAPS THAT THE LITE CLIENT IS NEW,
+		// AND THAT WE WILL SYNC TO AND VERIFY LATEST COMMIT.
+		fmt.Printf("trusting source at height %v and hash %X...\n", latestCommit.Height, latestCommit.Hash())
+		if options.Callback != nil {
+			err := options.Callback(latestCommit.Height, latestCommit.Hash())
+			if err != nil {
+				return types.SignedHeader{}, err
+			}
+		}
+		return latestCommit, nil
+	}
+}
+
+//----------------------------------------
+
+type nowFn func() time.Time
+
+const sizeOfPendingMap = 1024
+
+var _ lite.UpdatingProvider = (*Provider)(nil)
+
+// Provider implements a persistent caching Provider that
+// auto-validates.  It uses a "source" Provider to obtain the needed
+// FullCommits to securely sync with validator set changes.  It stores properly
+// validated data on the "trusted" local system.
+// NOTE: This Provider can only work with one chainID, provided upon
+// instantiation.
+type Provider struct {
+	chainID     string
+	logger      log.Logger
+	trustPeriod time.Duration // e.g. the unbonding period, or something smaller.
+	now         nowFn
+	height      int64
+
+	// Already validated, stored locally
+	trusted lite.PersistentProvider
+
+	// New info, like a node rpc, or other import method.
+	source lite.Provider
+
+	// pending map to synchronize concurrent verification requests
+	mtx                  sync.Mutex
+	pendingVerifications map[int64]chan struct{}
```

Comment:
I do not understand. The comments on `UpdateToHeight` say that you should use `concurrentProvider` if you want concurrent verification? Also both the map and its mutex aren't used anywhere. Please delete if you don't plan to use them.